[FEAT/#75] login / firebase authentication 도입

TaewoongR opened the pull request:
<details>
  <summary>[FEAT/#75]</summary>

- closed #75 

## *📍 Work Description*

- firebase authentication 관련 의존성 추가
- firebase authentication 도입하여 구글 소셜 로그인 구현
- 멀티 모듈 정리
- ui에 반영

## *📸 Screenshot*
(<img src="https://github.com/user-attachments/assets/6d24abe7-321c-42c4-979a-9139719dc309" width=270 />)


## *📢 To Reviewers*
- firebase authentication을 이용해서 구글 소셜 로그인을 구현했습니다.
- 처음 해보는 작업인지라 많은 우여곡절을 겪었습니다.
  - firebase 공식문서에는 xml 코드만 있다!!
  - 있는 코드도 Deprecate된 함수들을 사용한다...
  - 안드로이드 공식 문서에 해답이 있었다. 다만 사용법만 있어서 멀티모듈로 변환하는 작업이 필요했다.
  - hilt관련 문제로 5시간은 씨름한 것 같다. -> app 모듈에 의존성을 추가하여 해결!
- MVI로 바꾸는 작업을 추후에 진행하겠습니다! 
- 위의 스크린샷은 연출된 장면입니다! 이 브랜치를 받아서 돌리면 이 화면이 나오지 않습니다. 
- 버전별 대응을 해야할 것 같습니다.
- LocalProperty에 포함될 값 `비밀이야기`에 추가했습니다.

## ⏲️Time

    - 12 h +
</details>

<details>
  <summary>[FIX/#84]</summary>

- closed #84

## *📍 Work Description*
- data:auth, domain:auth -> core:auth 이전
- 의존성 설정
- google auth 와 firebase auth 를 통해 googleIdToken, UserInfo, firebaseUID 가져옴
- Activity Context를 사용하여 API30~33 에서 구글 로그인이 동작을 안하던 문제 해결

## *📸 Screenshot*

https://github.com/user-attachments/assets/a947ffa5-35db-4a20-8182-1f27db3ec011

<img src="https://github.com/user-attachments/assets/62f12e1c-9bc7-4e0c-b4bd-a9bb866e40bd" width=500>

## *📢 To Reviewers*
- 현재 로그인 화면에서는 어떠한 정보도 저장되지 않아 로그인 성공시 Toast 화면에 유저 정보만 나타나도록 했습니다.
- 구글 로그인 로직이 context와 얽혀있을 뿐만 아니라 Activity Context를 요구하기 때문에 hilt 는 제외했습니다.
- 동작 순차
  - 1. google cloud authentication 에서 유저 인증을 요청하고 Credential를 받습니다.
  - 2. Credential 이 올바른지 검사합니다. 올바르다면 GoogleIdTokenCredential 으로 타입캐스팅 됩니다.
  - 3. GoogleIdTokenCredential 의 idToken를 사용하여 firebase에 로그인을 요청합니다. 요청의 성공/실패 여부에 따라 개인 정보를 얻을 수 있습니다.
  - 4. 성공시 GoogleIdTokenCredential 에서 유저 정보(idToken 포함)와 firebaseUser의 UID를 뷰모델에 전달합니다.

- idToken : firebase 에 로그인 인증을 하기 위해 구글에서 발급한 토큰 (JWT)
- UID : firebase에서 유저를 구분하기 위해 발급된 시리얼(영구)

- 현재 구글 계정에서 임시로 정보를 추출했습니다.
  - 성, 이름
  - 휴대폰 번호
  - idToken
- 이것 외에도 아래와 같이 몇 가지 정보를 더 제공해줍니다.
<img src="https://github.com/user-attachments/assets/f1991b4f-a008-4f04-bded-fe334bd06378" width=500>

- local.properties 에 GOOGLE_WEB_CLIENT_ID 값을 별도로 추가해야 빌드가 됩니다. (local.properties gitignore 때문에 lint 빌드 실패)

## ⏲️Time

    - 5 h

</details>

#### closed #80
- Rebase and Merge conflict 후에 로컬에서 rebase 하여 새로운 PR 생성했습니다.

TaewoongR committed:
feat/#75 : 구글 oauth 의존성 추가

- com.google.android.gms:play-services-auth
- com.google.android.libraries.identity.googleid:googleid
- androidx.credentials:credentials

Co-authored-by: shinythinking <shinythink@gmail.com>

TaewoongR committed:
feat/#75 : 구글 아이콘 추가

Co-authored-by: shinythinking <shinythink@gmail.com>

TaewoongR committed:
feat/#75 : 임시 로그인 화면 생성

- 임시로 Material 화면으로 구성

Co-authored-by: shinythinking <shinythink@gmail.com>

TaewoongR committed:
feat/#75 : Google Cloud Oauth Web Client Id 추가

- local.property에 값을 저장하고 BuildConfig를 통해 얻도록 설정

Co-authored-by: shinythinking <shinythink@gmail.com>

TaewoongR committed:
feat/#75 : Google Oauth 로직 설정

- 뷰모델에서 로직 설계
- 이를 domain, data 모듈로 이전 예정

Co-authored-by: shinythinking <shinythink@gmail.com>

shinythinking committed:
chore/#75 : credential 의존성 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
chore/#75 : 도메인 모듈 의존성 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 도메인 auth, User 모델 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 도메인 auth, SignInWithGoogleUseCase 구현

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 도메인 auth, AuthRepository 구현
* 현재는 로그인 함수만 존재
* 로그아웃은 구현 예정

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
chore/#75 : 데이터 auth, 의존성 추가
* Local에서 Client에 대한 값을 받아오는 로직 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 데이터 auth, GoogleAuthDataSource 인터페이스 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 데이터 auth, GoogleAuthDataSource 구현체 추가
* application context를 사용해서 credential manager를 사용
* setNounce를 사용하여 보안 강화

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 데이터 auth, FirebaseAuthDataSource 인터페이스 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 데이터 auth, FirebaseAuthDataSource 구현체 추가
* firebaseAuth를 이용해서 인증 수행

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 데이터 auth, mapper 클래스 추가
* 도메인의 User와 AuthResult와 매핑

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 데이터 auth, AuthRepository 구현체 추가
* googleAuthDataSource와 firebaseAuthDataSource, UserMapper를 hilt로 받아 사용

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 데이터 auth, di 추가
* FirebaseAuth
* GoogleAuthDataSource
* UserMapper
* AuthRepository
* 의 의존성 주입!

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
chore/#75 : 피처 auth, 의존성 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
chore/#75 : 임시 화면(시험용) 삭제

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 피처 auth, AuthUiState 설정
* 성공시, User를 인수로 받도록 설정

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 피처 auth, AuthViewModel 구현
* Usecase를 받아서 구글 로그인 구현

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 피처 auth, di 설정
* SignInWithGoogleUseCase 의존성 주입!

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 피처 auth, LoginScreen 구글 로그인 버튼 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 피처 auth, LoginScreenRoute 추가
* Route에서 uiState 변화를 캐치

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
fix/#75 : ktlint의 요구사항...

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
fix/#75 : ktlint의 요구사항2...

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
refactor/#75 : bundle.auth로 의존성 묶기

shinythinking committed:
refactor/#75 : User 데이터 클래스를 core 모듈로 이동

TaewoongR committed:
fix/#84 : core:auth 모듈 생성

TaewoongR committed:
fix/#84 : google oauth 로직 위치 이전

TaewoongR committed:
fix/#84 : google oauth 결과 데이터 클래스 타입 선언

- 구글 계정에서 받은 실명과 휴대폰 번호
- 자동 로그인을 위한 google idToken
- firebase에서 받은 유저 구분을 위한 UID

TaewoongR committed:
fix/#84 : 모듈 삭제

- 로직 이전 후 필요 없는 모듈 일괄 삭제 및 의존성 수정

TaewoongR committed:
fix/#84 : 추가적인 미사용 파일 삭제

TaewoongR committed:
fix/#84 : 의존성 수정

- feature:login 에서 auth 관련 의존성 삭제

TaewoongR committed:
fix/#84 : core:auth 와 feature:login 연결

- GoogleOauth - ViewModel- Screen 연결
- Flow의 callbackFlow 내부에서 trySend 를 반환해야 collect 호출 시 emit

TaewoongR committed:
fix/#84 : fix for ktlint

TaewoongR committed:
fix/#84 : fix for ktlint

shinythinking reviewed (approved):
태웅님...ㅠㅠ 캐리해주셔서 감사합니다!

haeti-dev commented on core/auth/src/main/java/com/boostcamp/mapisode/auth/GoogleOauth.kt:
```diff
@@ -0,0 +1,109 @@
+			try {
+				val credentialManager = initializeCredentialManager()
+				val googleIdOption = createGoogleSignInOption()
+				val request = createCredentialRequest(googleIdOption)
+				Timber.e("Request: %s", request)
+
+				val resultCredential = credentialManager.getCredential(context, request).credential
+				Timber.e("Credential: %s", resultCredential)
```
팀버는 제거해주세요

TaewoongR replied:
네 적용하겠습니다!

haeti-dev commented on core/auth/src/main/java/com/boostcamp/mapisode/auth/GoogleOauth.kt:
```diff
@@ -0,0 +1,109 @@
+				val firebaseUID = authResult.user?.uid ?: throw Exception("Firebase UID가 없습니다.")
+				val googleIdToken = validatedCredential.idToken
+				val googleName = validatedCredential.run { "${familyName ?: ""}${givenName ?: ""}" }
+				val googlePhoneNumber = validatedCredential.phoneNumber ?: ""
```
전화번호 가져오는 거는 별도의 권한 요청없이 가져와도 되나요? 문제가 없는지 궁금합니다. 그리고 전화번호가 저희 프로젝트에 필요가 없는데 불러온 이유가 궁금합니다

TaewoongR replied:
그러네요. 개인정보는 법률에 영향을 받으니 함부로 사용할 수 없겠네요.
추가한 이유는 혹시나 필요하지 않을까 싶어서 가져왔습니다. 
이 부분도 수정하겠습니다.

haeti-dev commented on core/auth/src/main/java/com/boostcamp/mapisode/auth/LoginState.kt:
```diff
@@ -0,0 +1,8 @@
+	data class Error(val message: String) : LoginState
+}
+
+data class UserInfo(val firebaseUID: String, val name: String, val phoneNumber: String)
```
도메인에 있는 User 모델과 이 모델을 구분한 이유가 있나요?

TaewoongR replied:
받는 정보와 저장되는 정보가 다를거라 생각하고 구분했는데 생각해보니 다를 이유가 없겠어요.

haeti-dev commented on feature/login/build.gradle.kts:
```diff
@@ -5,3 +5,8 @@ plugins {
 }
+
+dependencies {
+	implementation(projects.core.model)
```
core:model은 이미 MapisodeFeaturePlugin에 등록되어 있습니다. 제거해주세요.

TaewoongR replied:
디테일 감사해요~

haeti-dev commented on feature/login/src/main/java/com/boostcamp/mapisode/login/AuthViewModel.kt:
```diff
@@ -0,0 +1,44 @@
+			GoogleOauth(context).googleSignIn().collect { result ->
+				_uiState.value = when (result) {
+					is LoginState.Success -> {
+						Timber.e("result: ${result.userInfo}")
```
로그 제거해주세요

TaewoongR replied:
팀버 고마웠어 안녕==~

haeti-dev reviewed (approved):
고생하셨습니다. 리뷰 남긴 건 사소한거라 먼저 승인할게요 꼭 고쳐주세요~

TaewoongR committed:
fix/#84 : 의존성 중복 제거

TaewoongR committed:
fix/#84 : core:model 의 User 데이터 클래스 수정 및 활용

- 유저 구글 계정의 휴대폰 번호 활용 제거

TaewoongR committed:
fix/#84 : timber 제거

Ameri-Kano reviewed (approved):
domain, data:auth 를 core:auth 로 옮기는 과정이군요! 고생 많으셨습니다!

TaewoongR committed:
fix/#84 : github pr_checker yml 수정

- local.properties GOOGLE_WEB_CLIENT_ID를 깃헙 빌드시 Repository Secrets 에 저장된 값을 곧바로 읽지 못하기 때문에 설정
